optimize/cmaes.go

```diff
@@ -0,0 +1,382 @@
+// but the covariance update equation is not identical.
+//
+// For more information about the CMA-ES algorithm, see
+//  https://en.wikipedia.org/wiki/CMA-ES
+//  https://arxiv.org/pdf/1604.00772.pdf
+type CmaEsChol struct {
+	// InitStepSize sets the initial size of the covariance matrix adaptation.
+	// If InitStepSize is 0, a default value of 0.5 is used. InitStepSize cannot
+	// be negative, or CmaEsChol will panic.
+	InitStepSize float64
+	// Population sets the population size for the algorithm. If Population is
+	// 0, a default value of 4 + math.Floor(3*math.Log(float64(dim))) is used.
+	// Population cannot be negative or CmaEsChol will panic.
+	Population int
+	// InitMean is the initial mean of the multivariate normal for sampling
+	// input locations. If InitMean is nil, the zero vector is used. If InitMean
+	// is not nil, it must have length equal to the problem dimension.
+	InitMean []float64
+	// InitCholesky specifies the Cholesky decomposition of the covariance
+	// matrix for the initial sampling distribution. If InitCholesky is nil,
+	// a default value of I is used. If it is non-nil, then it must have
+	// InitCholesky.Size() be equal to the problem dimension.
+	InitCholesky *mat.Cholesky
+	// StopLogDet sets the threshold for stopping the optimization if the
+	// distribution becomes too peaked. The log determinant is a measure of the
+	// (log) "volume" of the normal distribution, and when it is too small
+	// the samples are almost the same. If the log determinant of the covariance
+	// matrix becomes less than StopLogDet, the optimization run is concluded.
+	// If StopDeterminant is 0, a default value of dim*log(1e-16) is used.
+	// If StopDeterminant is NaN, the stopping criteria is not used, though
+	// this can cause numeric instabilities in the algorithm.
+	StopLogDet float64
+	// ForgetBest, when true, does not track the best overall function value found,
+	// instead returning the new best sample in each iteration. If ForgetBest
+	// is false, then the minimum value returned will be the lowest across all
+	// iterations, regardless of when that sample was generated.
+	ForgetBest bool
+	// Src allows a random number generator to be supplied for generating samples.
+	// If Src is nil the generator in math/rand is used.
+	Src *rand.Rand
+
+	// Fixed algorithm parameters.
+	dim                 int
+	pop                 int
+	weights             []float64
+	muEff               float64
+	cc, cs, c1, cmu, ds float64
+	eChi                float64
+
+	// Function data.
+	xs *mat.Dense
+	fs []float64
+
+	// Adaptive algorithm parameters.
+	invSigma float64 // inverse of the sigma parameter
+	pc, ps   []float64
+	mean     []float64
+	chol     mat.Cholesky
+
+	// Parallel fields.
+	taskIdxs []int          // Stores which simulation the task ran.
+	evals    []int          // remaining evaluations in this iteration.
+	mux      sync.Mutex     // protect access to evals.
```

Comment:
newline added (not sure if that's what you meant)

Reply:
By convention a mutex is placed in a stanza above the fields it protects. So this would be

```
	// Parallel fields.
	mu       sync.Mutex     // protect access to evals.
	taskIdxs []int // Stores which simulation the task ran.
	evals    []int // remaining evaluations in this iteration.

	wg sync.WaitGroup // wait for simulations to finish before iterating.
```

Note, not `mux` - this is overloaded with multiplex; either `mu` or `lock`.